pkg/streamingpromql/operators/aggregation.go

```diff
@@ -152,61 +156,100 @@ func (a *Aggregation) NextSeries(ctx context.Context) (types.InstantVectorSeries
 		thisSeriesGroup := a.remainingInnerSeriesToGroup[0]
 		a.remainingInnerSeriesToGroup = a.remainingInnerSeriesToGroup[1:]
 
-		if thisSeriesGroup.sums == nil {
+		if len(s.Floats) > 0 && thisSeriesGroup.floatSums == nil {
 			// First series for this group, populate it.
 
-			thisSeriesGroup.sums, err = a.Pool.GetFloatSlice(steps)
+			thisSeriesGroup.floatSums, err = a.Pool.GetFloatSlice(steps)
 			if err != nil {
 				return types.InstantVectorSeriesData{}, err
 			}
 
-			thisSeriesGroup.present, err = a.Pool.GetBoolSlice(steps)
+			thisSeriesGroup.floatPresent, err = a.Pool.GetBoolSlice(steps)
 			if err != nil {
 				return types.InstantVectorSeriesData{}, err
 			}
+			thisSeriesGroup.floatSums = thisSeriesGroup.floatSums[:steps]
+			thisSeriesGroup.floatPresent = thisSeriesGroup.floatPresent[:steps]
+		}
 
-			thisSeriesGroup.sums = thisSeriesGroup.sums[:steps]
-			thisSeriesGroup.present = thisSeriesGroup.present[:steps]
+		if len(s.Histograms) > 0 && thisSeriesGroup.histogramSums == nil {
+			// First series for this group, populate it.
+			thisSeriesGroup.histogramSums, err = a.Pool.GetHistogramPointerSlice(steps)
+			if err != nil {
+				return types.InstantVectorSeriesData{}, err
+			}
+			thisSeriesGroup.histogramSums = thisSeriesGroup.histogramSums[:steps]
 		}
 
 		for _, p := range s.Floats {
 			idx := (p.T - start) / interval
-			thisSeriesGroup.sums[idx] += p.F
-			thisSeriesGroup.present[idx] = true
+			thisSeriesGroup.floatSums[idx] += p.F
+			if !thisSeriesGroup.floatPresent[idx] {
+				thisSeriesGroup.floatPointCount++
+			}
+			thisSeriesGroup.floatPresent[idx] = true
+		}
+
+		for _, p := range s.Histograms {
+			idx := (p.T - start) / interval
+			if thisSeriesGroup.histogramSums[idx] == nil {
+				// We copy here because we modify the histogram through Add later on.
+				// It is necessary to preserve the original Histogram in case of any range-queries using lookback.
+				thisSeriesGroup.histogramSums[idx] = p.H.Copy()
+				thisSeriesGroup.histogramPointCount++
+			} else {
+				thisSeriesGroup.histogramSums[idx] = thisSeriesGroup.histogramSums[idx].Add(p.H)
+			}
 		}
```

Comment:
What if we do something like this? Then we only pay the price of copying the histogram when we need to, which should only be the case when lookback is involved.

```go
var lastUncopiedHistogram *histogram.FloatHistogram

for _, p := range s.Histograms {
	idx := (p.T - start) / interval
	if thisSeriesGroup.histogramSums[idx] != nil {
		thisSeriesGroup.histogramSums[idx] = thisSeriesGroup.histogramSums[idx].Add(p.H)
	} else if lastUncopiedHistogram == p.H {
		// We've already used this histogram for a previous point due to lookback.
		// Make a copy of it so we don't modify the other point.
		thisSeriesGroup.histogramSums[idx] = p.H.Copy()
		thisSeriesGroup.histogramPointCount++
	} else {
		// This is the first time we have seen this histogram.
		// It is safe to store it and modify it later without copying, as we'll make copies above if the same histogram is used for subsequent points.
		thisSeriesGroup.histogramSums[idx] = p.H
		thisSeriesGroup.histogramPointCount++
		lastUncopiedHistogram = p.H
	}
}
```

Reply:
I don't think that will work.

Because after the first time we have a point at this `idx` we add it to the histogram, if we see the histogram again (due to a lookback), it will have already been modified. So we can't just copy it at that point as it's already wrong. (Even if it was right, it would overwrite our summation, but we could work around that but that's not what's happening here).

To be sure, I quickly tested this and confirmed it does not pass the tests.

The challenge is that we need to be sure ahead of time that we won't see the same histogram again. Otherwise we can't confidentially modify it (or in other words we can't confidentially add to it). I'm not sure there is a sensible way to do this without implementing our own lookback. Even then, the challenge of keeping track of it would likely outweigh just making a copy at this point.

Reply:
I think I'm missing something sorry - what will modify `p.H` while we're iterating through a single series' samples and therefore modify a histogram before we've had a chance to copy it? 

I'm imagining we have two input series with histogram samples like this, including some instances of lookback:

| | t=0 | t=1 | t=2 | t=3 |
|--------|--------|--------|--------|--------|
| Series 1 | `h1` | `h1` | (nothing) | `h2` |
| Series 2 | `h3` | `h3` | `h4` | `h4` |

After we've processed series 1, with the change I suggested, then `histogramSums` will be:

| | t=0 | t=1 | t=2 | t=3 |
|--------|--------|--------|--------|--------|
| Output | `h1` | `Copy of h1` | (nothing) | `h2` |

And then after series 2:

| | t=0 | t=1 | t=2 | t=3 |
|--------|--------|--------|--------|--------|
| Output | `h1.Add(h3)` | `(Copy of h1).Add(h3)` | `h4` | `h2.Add(h4)` |

At no point do we mutate a histogram from a series before we've finished processing that series. For example, we only mutate `h1` when we start processing t=0 for series 2, but we've already copied it for t=1 by that point.